plugins/Actions/tests/Integration/ActionSiteSearchTest.php

```diff
@@ -0,0 +1,67 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link    https://matomo.org
+ * @license https://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\Actions\tests\Unit;
+
+use Piwik\Plugins\Actions\Actions\ActionSiteSearch;
+use Piwik\Tests\Framework\Fixture;
+use Piwik\Tests\Framework\TestCase\IntegrationTestCase;
+use Piwik\Tracker\Request;
+
+/**
+ * @group Actions
+ * @group SiteSearch
+ * @group Plugins
+ */
+class ActionSiteSearchTest extends IntegrationTestCase
+{
+    public function setUp()
+    {
+        parent::setUp();
+
+        Fixture::createWebsite('2010-01-01');
+    }
+
+    /**
+     * @dataProvider getSiteSearchData
+     */
+    public function testSiteSearchDetection($website, $url, $expectedResult)
+    {
+        $action = new ActionSiteSearchPublic(new Request(['idsite' => 1]));
+        $this->assertEquals($expectedResult, $action->detectSiteSearchFromUrl($website, parse_url($url)));
+    }
+
+    public function getSiteSearchData()
+    {
+        $defaultWebsite = [
+            'idSite' => 1,
+            'sitesearch_keyword_parameters' => ['q','k'],
+            'sitesearch_category_parameters' => ['cat','cc']
+        ];
+
+        return [
+            [$defaultWebsite, 'http://example.org/index.htm?q=keyword', ['http://example.org/index.htm', 'keyword', '', false]],
+            [$defaultWebsite, 'http://example.org/index.htm#q=keyword&cat=test', ['http://example.org/index.htm', 'keyword', 'test', false]],
+            [$defaultWebsite, 'http://example.org/index.htm#&q=keyword', ['http://example.org/index.htm', 'keyword', '', false]],
+            [$defaultWebsite, 'http://example.org/index.htm#?cat=test&q=keyword', ['http://example.org/index.htm', 'keyword', 'test', false]],
+            [$defaultWebsite, 'http://example.org/index.htm#?cat=test&q=keyword&otherparam=1', ['http://example.org/index.htm#otherparam=1', 'keyword', 'test', false]],
+            [$defaultWebsite, 'http://example.org/index.htm?cat=test&otherparam=1&q=keyword', ['http://example.org/index.htm?otherparam=1', 'keyword', 'test', false]],
+            [$defaultWebsite, 'http://example.org/index.htm#anchor?cat=test&otherparam=1&q=keyword', ['http://example.org/index.htm#anchor?otherparam=1', 'keyword', 'test', false]],
+            [$defaultWebsite, 'http://example.org/index.htm?cat=test&otherparam=1&q=kw#?q=keyword', ['http://example.org/index.htm?otherparam=1', 'keyword', 'test', false]],
+            [$defaultWebsite, 'http://example.org/index.htm?k=keyword&cc=cat', ['http://example.org/index.htm', 'keyword', 'cat', false]],
```

Comment:
Some other test cases that could be added:

- `http://example.org/index.html?a=b#?&&&?q=keyword`
- `http://example.org/#&?q=keyword`
- `#?q=keyword`
- `http://example.org/#&?var=val?q=keyword`
- `http://example.org/#q=keyword?q=anotherkeyword`
- maybe some other weird combos

What do you think? They're not likely to occur in practice, but it might be good to define the behavior in these cases.

Reply:
sure. why not